netket/vqs/mc/mc_state/state.py

```diff
@@ -293,38 +294,50 @@ def sampler(self, sampler: Sampler):
                 )
             )
 
+        # Save the old `n_samples` before the new `sampler` is set.
+        # `_chain_length == 0` means that this `MCState` is being constructed.
+        if self._chain_length > 0:
+            n_samples_old = self.n_samples
+
         self._sampler = sampler
         self.sampler_state = self.sampler.init_state(
             self.model, self.variables, seed=self._sampler_seed
         )
+
+        # Update `n_samples`, `n_samples_per_rank`, and `chain_length` according
+        # to the new `sampler.n_chains`.
+        # If `n_samples` is divisible by the new `sampler.n_chains`, it will be
+        # unchanged; otherwise it will be rounded up.
+        # If the new `n_samples_per_rank` is not divisible by `chunk_size`, a
+        # `ValueError` will be raised.
+        # `_chain_length == 0` means that this `MCState` is being constructed.
+        if self._chain_length > 0:
+            self.n_samples = n_samples_old
+
         self.reset()
 
     @property
     def n_samples(self) -> int:
         """The total number of samples generated at every sampling step."""
-        return self._n_samples
+        return self.chain_length * self.sampler.n_chains
 
     @n_samples.setter
     def n_samples(self, n_samples: int):
         chain_length = compute_chain_length(self.sampler.n_chains, n_samples)
 
-        n_samples_per_rank = chain_length * self.sampler.n_chains_per_rank
         n_samples = chain_length * self.sampler.n_chains
-
         check_chunk_size(n_samples, self.chunk_size)
 
-        self._n_samples = n_samples
         self._chain_length = chain_length
-        self._n_samples_per_rank = n_samples_per_rank
         self.reset()
```

Comment:
Since you changed the 'fundamental' unit, can you change this code to be 
```python
    @n_samples.setter
     def n_samples(self, n_samples: int):
         chain_length = compute_chain_length(self.sampler.n_chains, n_samples)
         self.chain_length = chain_length
```

and modify the chain length setter to be 
```python
def chain_lengthsetter(self, chain_length):
    n_samples = chain_length * self.sampler.n_chains
    check_chunk_size(n_samples, self.chunk_size)
    self._chain_length = chain_length
    self.reset()
```

I think it will make more sense

Reply:
There should be a check that `chain_length` should be positive, and there is a test for that. Let me also check this in `chain_length.setter`